src/ops/reduction.rs

```diff
@@ -0,0 +1,298 @@
+        keep_dims: impl Into<Option<bool>>,
+        stream: StreamOrDevice,
+    ) -> Array {
+        self.try_all_axes_device(&[axis], keep_dims, stream)
+            .unwrap()
+    }
+
+    /// An `and` reduction over the given axis without validating axis is valid for the array.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], &[3, 4]);
+    /// let mut b = unsafe { a.all_axis_unchecked(0, None) };
+    ///
+    /// b.eval();
+    /// let results: &[bool] = b.as_slice();
+    /// // results == [false, true, true, true]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - axis: The axis to reduce over
+    /// - keep_dims: Whether to keep the reduced dimensions -- defaults to false if not provided
+    /// - stream: The stream to execute the operation on
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not validate that the axis is valid for the array.
+    #[default_device]
+    pub unsafe fn all_axis_device_unchecked(
+        &self,
+        axis: i32,
+        keep_dims: impl Into<Option<bool>>,
+        stream: StreamOrDevice,
+    ) -> Array {
+        self.all_axes_device_unchecked(&[axis], keep_dims, stream)
+    }
+
+    /// An `and` reduction over the given axis returning an error if the axis is invalid.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], &[3, 4]);
+    /// let mut b = a.try_all_axis(0, None).unwrap();
+    ///
+    /// b.eval();
+    /// let results: &[bool] = b.as_slice();
+    /// // results == [false, true, true, true]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - axis: The axis to reduce over
+    /// - keep_dims: Whether to keep the reduced dimensions -- defaults to false if not provided
+    /// - stream: The stream to execute the operation on
+    #[default_device]
+    pub fn try_all_axis_device(
+        &self,
+        axis: i32,
+        keep_dims: impl Into<Option<bool>>,
+        stream: StreamOrDevice,
+    ) -> Result<Array, OperationError> {
+        self.try_all_axes_device(&[axis], keep_dims, stream)
+    }
+
+    /// An `and` reduction over all axes.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], &[3, 4]);
+    /// let mut b = a.all(None);
+    ///
+    /// b.eval();
+    /// let results: &[bool] = b.as_slice();
+    /// // results == [false]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - keep_dims: Whether to keep the reduced dimensions -- defaults to false if not provided
+    /// - stream: The stream to execute the operation on
+    #[default_device]
+    pub fn all_device(&self, keep_dims: impl Into<Option<bool>>, stream: StreamOrDevice) -> Array {
+        let axes: Vec<i32> = (0..self.ndim() as i32).collect();
```

Comment:
@minghuaw know a better way to do this one?

Reply:
Should we just merge `all_axes_*`, `all_axis_*` into `all_*`? Since that's how we are dealing with optional params in other methods?